crates/kitsune-core/src/activitypub/fetcher.rs

```diff
@@ -727,6 +756,39 @@ mod test {
         .await;
     }
 
+    #[tokio::test]
+    #[serial_test::serial]
+    async fn check_ap_content_type() {
+        database_test(|db_pool| async move {
+            let client = service_fn(|_: Request<_>| async {
+                Ok::<_, Infallible>(Response::new(Body::empty()))
+            });
+            let client = Client::builder().service(client);
+
+            let fetcher = Fetcher::builder()
+                .client(client.clone())
+                .db_pool(db_pool)
+                .embed_client(None)
+                .federation_filter(
+                    FederationFilterService::new(&FederationFilterConfiguration::Deny {
+                        domains: Vec::new(),
+                    })
+                    .unwrap(),
+                )
+                .search_service(NoopSearchService)
+                .webfinger(Webfinger::with_client(client, Arc::new(NoopCache.into())))
+                .post_cache(Arc::new(NoopCache.into()))
+                .user_cache(Arc::new(NoopCache.into()))
+                .build();
+
+            assert!(matches!(
+                fetcher.fetch_object("https://example.com/fakeobject").await,
+                Err(Error::Api(ApiError::BadRequest))
+            ));
```

Comment:
The test would yield an error even without the new content type check, so it doesn't reliably distinguish the intended error from errors caused by other random reasons. I'd reuse the `handler` helper function and override its content type to simulate a response which is the same as a legitimate one except the content type:

```suggestion
            let client = service_fn(|req: Request<_>| async {
                let mut res = handle(req).unwrap();
                res.headers_mut().remove(CONTENT_TYPE);
                Ok::<_, Infallible>(res)
            });
            let client = Client::builder().service(client);

            let fetcher = Fetcher::builder()
                .client(client.clone())
                .db_pool(db_pool)
                .embed_client(None)
                .federation_filter(
                    FederationFilterService::new(&FederationFilterConfiguration::Deny {
                        domains: Vec::new(),
                    })
                    .unwrap(),
                )
                .search_service(NoopSearchService)
                .webfinger(Webfinger::with_client(client, Arc::new(NoopCache.into())))
                .post_cache(Arc::new(NoopCache.into()))
                .user_cache(Arc::new(NoopCache.into()))
                .build();

            assert!(matches!(
                fetcher.fetch_object("https://corteximplant.com/users/0x0/statuses/109501659207519785").await,
                Err(Error::Api(ApiError::BadRequest))
            ));
```